src/realm.js

```diff
@@ -308,6 +308,37 @@ export class Realm {
     return context;
   }
 
+  // Puts the new LexicalEnvironment onto the current execution context (or the one specified)
+  pushScope(environment: LexicalEnvironment, context?: ExecutionContext) {
+    let executionContext = context ? context : this.getRunningContext();
+    let prevScope = executionContext.lexicalEnvironment;
+    executionContext.lexicalEnvironment = environment;
+    return prevScope;
+  }
+
+  // Destroys the LexicalEnvironment (scope) of the specified context (or the current execution context's)
+  popScope(prevEnvironment: LexicalEnvironment, context?: ExecutionContext) {
+    let execContext = context ? context : this.getRunningContext();
+    this.destroyScope(execContext);
+    if (execContext.lexicalEnvironment.parent !== null) invariant(execContext.lexicalEnvironment !== prevEnvironment);
+    execContext.lexicalEnvironment = prevEnvironment;
+  }
+
+  // Destroys a scope that may not be the running context's scope
+  // Clears the Bindings corresponding to the disappearing Scope from ModifiedBindings
+  destroyScope(context: ExecutionContext) {
```

Comment:
Why is the global scope special?

Reply:
The arguments a bit weird. Why does `pushScope` and `popScope` take a `LexicalEnvironment` as it's first argument, while `destroyScope` doesn't? Should it take one, and assert that it's equal to `context.lexicalEnvironment`?